Generate test coverage report in html

FedericoPonzi opened the pull request:
I wanted to start contributing some tests, but it was hard to understand which lines are not covered. I know there is a test-dist command, but for some reason Intellij was not able to load it and the generated report is also empty. In this PR, I'm adding a new command for ant `jacoco-report` that generates the HTML report. Using `ant -f customBuild.xml test-dist jacoco-report` will run the tests with 8 threads and generate the html report in `target/site/index.html`. If you're curious, the reported coverage is at 60% at the moment (according to this output, not sure how precise it is).
![image](https://github.com/tlaplus/tlaplus/assets/2124904/4890442b-a70f-42d4-a118-e3a8fe60cda5)

lemmy commented on tlatools/org.lamport.tlatools/customBuild.xml:
```diff
@@ -950,7 +950,7 @@
 		<mkdir dir="${test.reports}/onJar" />
 		<jacoco:coverage destfile="target/code-coverage.exec" includes="pcal.*:tla2sany.*:tla2tex.*:tlc2.*:util.*:org.lamport.*" excludes="com.sun.*:javax.*:**/Tests.*:**/*Test*">
-			<junit printsummary="yes" haltonfailure="${test.halt}" haltonerror="${test.halt}" forkmode="perTest" fork="yes">
+			<junit printsummary="yes" haltonfailure="${test.halt}" haltonerror="${test.halt}" forkmode="perTest" fork="yes" threads="8">
```
Should this really be hardcoded to 8 threads?

lemmy commented:
Beyond the Eclipse IDE, we used to visualize coverage either on our defunct Jenkins build machine or on SonarQube. As far as I know, the latter is no longer free. Could the HTML alternatively be generated and saved as part of our GitHub Action to keep the Ant build simple?

lemmy commented:
This PR can be build on https://github.com/tlaplus/tlaplus/pull/954, which adds the capability to upload certain files in `tlatools/org.lamport.tlatools/target/`. However, note that coverage data can only be collected by the `test-dist` target. This target should not be invoked in the `pr.yml` file. Additionally, the `test` Ant target workflow appears to be incompatible with parallel test execution.

lemmy commented:
Running this change locally fails with:

```shell
BUILD FAILED
/Users/markus/src/TLA/tla/tlatools/org.lamport.tlatools/customBuild.xml:1093: Problem: failed to create task or type target
Cause: The name is undefined.
Action: Check the spelling.
Action: Check that any custom tasks/types have been declared.
Action: Check that any <presetdef>/<macrodef> declarations have taken place.
```

FedericoPonzi commented:
> This PR can be build on #954, which adds the capability to upload certain files in `tlatools/org.lamport.tlatools/target/`. However, note that coverage data can only be collected by the `test-dist` target. This target should not be invoked in the `pr.yml` file. Additionally, the `test` Ant target workflow appears to be incompatible with parallel test execution.

If we don't want to run it per pr, what are the acceptance criteria for this one then? 

> Running this change locally fails with:
> 
> ```shell
> BUILD FAILED
> /Users/markus/src/TLA/tla/tlatools/org.lamport.tlatools/customBuild.xml:1093: Problem: failed to create task or type target
> Cause: The name is undefined.
> Action: Check the spelling.
> Action: Check that any custom tasks/types have been declared.
> Action: Check that any <presetdef>/<macrodef> declarations have taken place.
> ```

I've refactored it and rebased on master, can you please try again when you have a moment?

lemmy commented:
@FedericoPonzi This worked on my m1 after running the `test-dist` target:

```shell
markus@avocado [10:45:06] [~/Desktop/tlaplus/tlatools/org.lamport.tlatools] [html-coverage]
-> % ant -f customBuild.xml code-coverage-report
Buildfile: /Users/markus/Desktop/tlaplus/tlatools/org.lamport.tlatools/customBuild.xml

code-coverage-report:
    [mkdir] Created dir: /Users/markus/Desktop/tlaplus/tlatools/org.lamport.tlatools/target/code-coverage-report
[jacoco:report] Loading execution data file /Users/markus/Desktop/tlaplus/tlatools/org.lamport.tlatools/target/code-coverage.exec
[jacoco:report] Writing bundle 'TLA+ tools' with 1601 classes
     [echo] HTML report available here: target/code-coverage-report/index.html

BUILD SUCCESSFUL
Total time: 3 seconds
```

Consider adding it to the ant invocation and upload of workflow artifacts in main.yml and py.yml.

lemmy reviewed (approved):
LGTM, but please remove the unrelated formatting changes.

FedericoPonzi commented:
Not sure how they crept in there, I've included the minimal diffset. I'll create a followup pr for the upload of artifacts

lemmy commented:
To avoid merge conflicts, please be aware of https://github.com/tlaplus/tlaplus/pull/967.
